node/src/actors/chain_manager/handlers.rs

```diff
@@ -309,6 +311,33 @@ impl Handler<AddBlocks> for ChainManager {
                 let sync_target = self
                     .sync_target
                     .expect("The sync target should be defined for synchronizing");
+
+                let sync_superblock =
+                    self.sync_superblock
+                        .as_ref()
+                        .and_then(|(hash, superblock)| {
+                            if hash == &sync_target.superblock.hash_prev_block {
+                                Some(superblock.clone())
+                            } else {
+                                None
+                            }
+                        });
+
+                let sync_superblock = if let Some(sync_superblock) = sync_superblock {
+                    Some(sync_superblock)
+                } else if sync_target.superblock.checkpoint == 0 {
+                    None
+                } else {
+                    log::debug!("Received blocks before superblock");
+                    // Received the `AddBlocks` message before the `AddSuperBlock` message.
+                    // We cannot finish the synchronization without the sync_superblock, so in that
+                    // case, ask for the sync_superblock again, and try to handle the `AddBlocks`
+                    // message later
+                    self.request_sync_target_superblock(ctx, sync_target.superblock);
+                    ctx.notify_later(msg, Duration::from_secs(6));
+                    return;
+                };
```

Comment:
This logic is a bit hard to understand, I would rewrite it like this: (should be equivalent, but not sure)

```suggestion
                if sync_superblock.is_none() && sync_target.superblock.checkpoint != 0 {
                    log::debug!("Received blocks before superblock");
                    // Received the `AddBlocks` message before the `AddSuperBlock` message.
                    // We cannot finish the synchronization without the sync_superblock, so in that
                    // case, ask for the sync_superblock again, and try to handle the `AddBlocks`
                    // message later
                    self.request_sync_target_superblock(ctx, sync_target.superblock);
                    ctx.notify_later(msg, Duration::from_secs(6));
                    return;
                }
```